### DISCUSSION TITLE
PMD Fundaments: Up the Tree or Down the Tree?

### DISCUSSION THREAD

**Discussion by justinstroud-campmeeker** [pmd/pmd#4932] [Q&A]
I almost want to use a "cat in a tree" analogy here after reading my title.

I'm trying to find instances of counterfeit test coverage. The most common anti-pattern for this is a bunch of increment statements (i++ etc.) For "reasons" I'm doing this in a Java rule.

Is it more efficient to start at the atomic level (the statement visitor) or the method level (the method visitor)?

Better to let visit get hit a bunch of times and walk up the tree, or less hits to visit and walk down the tree? Which is most efficient within the domain of PMD?

**Comment by justinstroud-campmeeker** [comment]
There is the option to shut down the traversal during the statement vist (just not hit the super.) Right?

**Reply by jsotuyod** [reply]
Yes. Not calling super will prevent all descendant nodes from being inspected (assuming you are not using a call chain visitor, which will only visit the selected nodes, irrespective of super or not)

**Comment by jsotuyod** [comment]
PMD is REALLY efficient and will visit all nodes by default, as long as you are not doing anything silly it should not be an issue at all. This seems like early optimization.

I'd advise you get something to working, and then ask for insight if performance is an issue.

**Comment by oowekyala** [comment]
Juan's advice about early optimization is valid.

FWIW to get the best results you need to consider two things in order:
1. Correctness / simplicity of the rule. For your use case, since you want to check several statements, you are probably better off going top-down on an entire method. You should start writing your rule like that until it works.
2. Performance. Here what matters most is visiting as few nodes as possible. Often in top-down rules you have no choice but to explore the entire tree (or all method declarations) to find what you're looking for, because there is no more specific node your rule could anchor to. The best performing rules are those that anchor to a relatively rare node, for instance an empty statement, or a lambda expression. A method call node for instance is not rare at all, and there are many more of them than method declarations.

In your case, there might be a way to traverse fewer nodes than just all method declarations. You could traverse only those method declarations that contain at least an increment operation. This would save work because the increment statements should be rarer than method declarations in general (there are many method declarations that don't contain any increment).

So you could make your rule extend `AbstractJavaRuleChainRule` and target the node class `ASTUnaryExpression`. Then you check that the operator of the unary expression is an increment or decrement before proceeding (there are other unary ops like minus), and maybe also check that the parent is a statement, and not another expression. Then you go up to the containing method and you start traversing top down from there, with the logic you have already written. But you have to make sure that you don't explore a method twice if it contains two increment statements, maybe by setting a flag on the user data of the method ([getUserMap()](https://docs.pmd-code.org/apidocs/pmd-core/7.0.0/net/sourceforge/pmd/lang/ast/Node.html#getUserMap())). Good luck!

**Reply by justinstroud-campmeeker** [reply]
I don't even care what kind of mathematical operation is performed. I'm hard pressed to imagine a scenario where a method would have more than ten discreet operations like that. At that point, the method length itself is suspect. If I start at the method, I just need to count the instances of those operations and have a configurable maximum as a property. Done. I'll let PMD continue to check further methods, but once a method has been visited it shouldn't be visited again, right?

**Reply by jsotuyod** [reply]
Correct, PMD will visit each node exactly once as long as you don't mess with super / visit calls.
